src/nbl/asset/utils/CIESProfile.cpp

```diff
@@ -0,0 +1,300 @@
+    const size_t bufferRowLength = asset::IImageAssetHandlerBase::calcPitchInBlocks(width, texelBytesz);
+    region.bufferRowLength = bufferRowLength;
+    region.imageExtent = imgInfo.extent;
+    region.imageSubresource.baseArrayLayer = 0u;
+    region.imageSubresource.layerCount = 1u;
+    region.imageSubresource.mipLevel = 0u;
+    region.bufferImageHeight = 0u;
+    region.bufferOffset = 0u;
+
+    auto buffer = core::make_smart_refctd_ptr<asset::ICPUBuffer>(texelBytesz * bufferRowLength * height);
+
+    if (!outImg->setBufferAndRegions(std::move(buffer), core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<asset::IImage::SBufferCopy>>(1ull, region)))
+        return {};
+
+    //! Generate 2D IES grid data CandelaPower Distribution Curve texture can be created from
+    {
+        const auto& creationParams = outImg->getCreationParameters();
+
+        CFillImageFilter::state_type state;
+        state.outImage = outImg.get();
+        state.subresource.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+        state.subresource.baseArrayLayer = 0u;
+        state.subresource.layerCount = 1u;
+        state.outRange.extent = creationParams.extent;
+
+        const IImageFilter::IState::ColorValue::WriteMemoryInfo wInfo(creationParams.format, outImg->getBuffer()->getPointer());
+
+        const double maxValue = getMaxValue();
+        const double maxValueRecip = 1.0 / maxValue;
+
+        const double vertInv = 1.0 / height;
+        const double horiInv = 1.0 / width;
+
+        const double dTheta = core::PI<double>() / height;
+        const double dPhi = 2 * core::PI<double>() / width;
+
+        integral = 0.0;
+        size_t nonZeroEmissionDomainSize = 0u;
+
+        std::atomic<IES_STORAGE_FORMAT> aIntegral = 0.0;
+        std::atomic<size_t> aNonZeroEmissionDomainSize = 0;
+
+        auto fill = [&](uint32_t blockArrayOffset, core::vectorSIMDu32 position) -> void
+        {
+            const auto dir = octahdronUVToDir(((float)position.x + 0.5) * vertInv, ((float)position.y + 0.5) * horiInv);
+            const auto [theta, phi] = sphericalDirToRadians(dir);
+            const auto& intensity = sample(theta, phi);
+            const auto& value = intensity * maxValueRecip;
+
+            const decltype(integral) integrationV = dPhi * dTheta * std::sin(theta) * intensity;
+            
+            if (integrationV != 0.0)
+                if constexpr (policy._Parallelize)
+                {
+                    atomic_fetch_add<IES_STORAGE_FORMAT>(&aIntegral, integrationV);
+                    aNonZeroEmissionDomainSize.fetch_add(1u);
+                }
+                else
+                {
+                    integral += integrationV;
+                    ++nonZeroEmissionDomainSize;
+                }
+
+            const uint16_t encodeV = static_cast<uint16_t>(std::clamp(value * UI16_MAX_D, 0.0, UI16_MAX_D));
```

Comment:
remember about adding 0.5 to the value before mult by `UI16_MAX_D`

https://github.com/Devsh-Graphics-Programming/Nabla/pull/630#discussion_r1468267196

Reply:
also made it use asset:: encode

Reply:
>here it will break it imo
see now that you use `asset::encode` we can fix this later XD

So mark as resolved after reading this message.

Because they're all broken :kek:
https://github.com/Devsh-Graphics-Programming/Nabla/blob/f0796aa799b3889ba96f7792c105c933bee0d98a/include/nbl/asset/format/encodePixels.h#L628

Basically if you don't add 0.5, you get this:
```
0.0 -> 0.0 -> 0
0.999... -> 65 534.999934465 -> 65 535
1.0 -> 65 535.0 -> 65 535
```
so 0 gets mapped to too frequently, and 65 535 almost never

This doesn't get fixed with even with a dither of +/-0.5 because anything below 0.5/(2^16-1) will get mapped to 0, 100% of the time

Reply:
yeah I get what you mean here and why it's needed in our encodes but I'm not sure if you got what I mean at beginning

> about adding 0.5 to the **value**

so let's say I give you max intensity and value is 1.0f, you add 0.5f before multiplication and finally you end up with 1.5*65535 encode which is out of `uint16_t` scope, that's what I mean, it's just the addition would not be to the `value`

Reply:
this would also mean we map 0.0f -> 0.5f -> uint16_t(65535/2) which definitely should not take place

Reply:
ah yea I see I wrote "before" mistakenly. my bad